### PR TITLE
Header Length for improving parsing

### PR DESCRIPTION
**Pull request by suhasHere**
Addresses #474

### CODE REVIEW

**Review by afrind — COMMENTED** [review]
Individual Review: 

We also need to specify what happens when:

1) the section ends before the section length
2) the section extends beyond the section length (eg: the varint length would extend past the section length).

Probably ought to make both a protocol violation and tear the whole session down?

**Comment by vasilvv** [comment]
I think this is good for control messages, but I'd mildly prefer to keep the data stream headers without length, since those are sensitive to the framing overhead.

File: draft-ietf-moq-transport.md [hunk 1]
```diff
@@ -1760,13 +1760,17 @@ An `OBJECT_DATAGRAM` message carries a single object in a datagram.
...
 payload and send the Object as datagram. In certain scenarios where the object
 size can be larger than maximum datagram size for the session, the Object
 will be dropped.
 
+The `Object Header Length` specifies the total encoded length in bytes for
+all the fields except the Object Payload Length and the Object Payload.
+
 ~~~
 OBJECT_DATAGRAM Message {
+  Object Header Length (i),
```
[the remark this answers is not in the document]
Comment (TimEvens) [reply draft-ietf-moq-transport.md:1773]: +1

**Comment by TimEvens** [comment]
> I think this is good for control messages, but I'd mildly prefer to keep the data stream headers without length, since those are sensitive to the framing overhead.

Suppose it depends how often streams are changing. Adding object header length does add overhead, but it should be one byte in most uses.

**Comment by fluffy** [comment]
Victors comment is really making me think about if we need this on objects or not. I'm very torn and think we should talk about how to do while still keeping the framing overhead as low as possible.

**Comment by afrind** [comment]
Chair Comment:

It seems like there's general support for control messages, but some discussion required for objects.  Can we move the object changes to a separate PR so we can land the control changes for the next draft cut?

File: draft-ietf-moq-transport.md [hunk 2]
```diff
@@ -1797,10 +1801,13 @@ TODO: figure out how a relay closes these streams
...
 belong to the track requested in the Subscribe message identified by `Subscribe
 ID`.  All objects on the stream have the `Publisher Priority` specified in the
-stream header.
+stream header. The `Stream Header Length` specifies the total encoded length
+of the header in bytes which includes Subscribe ID, Track Alias and Publisher
+Priority fields.
 
 ~~~
 STREAM_HEADER_TRACK Message {
+  Stream Header Length (i),
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:1810]: I wouldn't bother fixing STREAM_HEADER_TRACK, since I expect it'll be removed soonish.

**Review by ianswett — COMMENTED** [review]
Adding a length for control messages seems sensible, but in other cases I don't think we need an additional length.

File: draft-ietf-moq-transport.md [hunk 3]
```diff
@@ -1877,17 +1893,20 @@ Sending a track on one stream:
 
 ~~~
 STREAM_HEADER_TRACK {
+  Stream Header Length = 3
```
[the remark this answers is not in the document]
Comment (fluffy) [reply draft-ietf-moq-transport.md:1903]: Trying to think about how I would code this either way. Lets say there were 62 bytes of data after this header. If I encode this number as 1 byte varint with value 63. Or I could encode as 2 byte varint with value of 64. Not sure how this changes what I think we should do - just an edge case worth thinking about.

**Pull request merged**